site/content/api/Device-Connection-API.md

```diff
@@ -0,0 +1,138 @@
+## Preconditions for invoking the Device Connection API
+
+1. Client has established an AMQP connection with the Device Connection service.
+1. Client has established an AMQP link in role *sender* on the connection using target address `device_con/${tenant_id}`. This link is used by the client to send commands concerning device connections to Hono.
+1. Client has established an AMQP link in role *receiver* on the connection using source address `device_con/${tenant_id}/${reply-to}` where *reply-to* may be any arbitrary string chosen by the client. This link is used by the client to receive responses to the requests it has sent to the Device Connection service. This link's source address is also referred to as the *reply-to* address for the request messages.
+
+{{< figure src="../connectToDeviceConnection.png" alt="A client establishes an AMQP connection and the links required to invoke operations of the Device Connection service" title="Client connecting to Device Connection service" >}}
+
+## Operations
+
+The operations described in the following sections are invoked by Hono's components during run time and are therefore mandatory to implement.
+
+### Set Last Known Gateway for Device
+
+Clients use this command to *set* the gateway that last acted on behalf of a given device.
+
+**Message Flow**
+
+
+The following sequence diagram illustrates the flow of messages involved in a *Client* setting the last known gateway for a device.
+
+![Set Last Known Gateway for Device message flow](../set-last-gw_Success.png)
+
+
+**Request Message Format**
+
+The following table provides an overview of the properties a client needs to set on a message to set the last known gateway for a device in addition to the [Standard Request Properties]({{< relref "#standard-request-properties" >}}).
+
+| Name         | Mandatory | Location                 | AMQP Type | Description |
+| :----------- | :-------: | :----------------------- | :-------- | :---------- |
+| *subject*    | yes       | *properties*             | *string*  | MUST be set to `set-last-gw`. |
+| *gateway_id* | yes       | *application-properties* | *string*  | The identifier of the gateway that last acted on behalf of the device identified by the *device_id* property. If a device connects directly instead of through a gateway, the device's identifier MUST be specified here. |
+
+The body of the message SHOULD be empty and will be ignored if it is not.
+
+**Response Message Format**
+
+A response to a *set last known gateway for device* request contains the [Standard Response Properties]({{< relref "#standard-response-properties" >}}).
+
+The response message's *status* property may contain the following codes:
+
+| Code | Description |
+| :--- | :---------- |
+| *204* | OK, the last known gateway for the device was updated. |
+| *404* | Not Found, there is no device registered with the given *device_id* within the given *tenant_id*, or no gateway registered with the given *gateway_id*. |
```

Comment:
returning a 404 would actually require the service implementation to check with the registration service if the device/gw exists. This, however, introduces a new dependency on the registration service which we wanted to eliminate with the introduction of this service in the first place ...
So we will probably end up accepting all identifiers, right?

Reply:
Right, that is still a leftover from defining the method in the registration API. I've removed both such places in the doc.

Reply:
FMPOV it is ok to remove the 404 and indeed *not* perform a check for existence in this service. However, I think we should explicitly mention that this means that every client can submit *any* data (within its tenant) and that implementors should therefore take care to e.g. restrict the overall amount of data that can be stored per tenant, in particular if the data is to be kept in memory only.

Reply:
Good point. I've added corresponding notes in the doc.